cdci_data_analysis/analysis/renku_helper.py

```diff
@@ -173,11 +170,11 @@ def create_new_notebook_with_code(repo, api_code, job_id, file_name=None):
 def commit_and_push_file(repo, file_path):
     try:
         add_info = repo.index.add(file_path)
-        commit_info = repo.index.commit("commit code from MMODA")
+        commit_info = repo.index.commit("commit code from MMODA " + time.strftime('%Y-%m-%dT%H:%M:%S'))
```

Comment:
normally, time is always already in the commit. So this is duplication.
What's less obvious is user. Maybe we can create commit authored by the user from the token of the request? The commiter will be still set from the config, ODA user.

Reply:
I wanted to have something generating a different commit message every time, otherwise the commit history would not change, especially now that we force-push always. 

The user is definitely an option, but perhaps I would also add a rand short string.

Reply:
the commit history for each branch should consist of a single ODA commit. Does it not?

Reply:
Ah ok I thought it would have been useful to see the history of the commits.

But then we should not put anything and leave the message as it is (just `commit code from MMODA`) otherwise two requests from two different users might cause the commit history for a branch to have more than one commit.

Reply:
Wait, which history? Every new branch should start from master, and have exactly one MMODA commit.

Reply:
> Wait, which history? 

The commit history for a new branch

> Every new branch should start from master, and have exactly one MMODA commit.

It does, but if we change the commit message including the author, two force-push on the same branch create two MMODA message.

Reply:
force-push rewrites history. So the resulting branch will be that of the last made request, but only one. The history of the branch before force-push will be lost - so there is never two messages in the same branch.

Reply:
> force-push rewrites history. 

Ok I was not considering this aspect: it rewrites the history on involved files (which for us is the notebook with the api_code, [example](https://renkulab.io/gitlab/gabriele.barni/test-dispatcher-endpoint/-/commits/2fcaf6e112729a8c0b184632d24464ed01f7847d/api_code_9a5e3646291b26da.ipynb) ), but for the branch will be still visible ( [example](https://renkulab.io/gitlab/gabriele.barni/test-dispatcher-endpoint/-/commits/mmoda_request_9a5e3646291b26da) )

Does it make sense or am I missing something?

Reply:
It is now including the username extracted from the token, if this is available (which should always be the case)

Reply:
so, is the user name, extracted from the token, in the commit message ok?

Reply:
> > force-push rewrites history.
> 
> Ok I was not considering this aspect: it rewrites the history on involved files (which for us is the notebook with the api_code, [example](https://renkulab.io/gitlab/gabriele.barni/test-dispatcher-endpoint/-/commits/2fcaf6e112729a8c0b184632d24464ed01f7847d/api_code_9a5e3646291b26da.ipynb) ), but for the branch will be still visible ( [example](https://renkulab.io/gitlab/gabriele.barni/test-dispatcher-endpoint/-/commits/mmoda_request_9a5e3646291b26da) )
> 
> Does it make sense or am I missing something?

I think there is some confusion about what is rewritten with what.

force push rewrites commit history (including all files, whether some files were added not in first commit is circumstantial) on the remote.
it overwrites the remote history with local history.
these mmoda branches are started from master.
in your example you made new commit on master.
so new branches have this new commit too.

what was in this branch on remote is completely discarded, removing all history.

Reply:
> so, is the user name, extracted from the token, in the commit message ok?

I rather meant to set commit author, not so much the message. Both commit author and timestamp are part  of the commit metadata, and are usually shown, e.g. in gitlab

Reply:
what could be more useful in the message, is the sort of metadata which is not naturally in the commit.

For example, it could say 
"stored API code of MMODA request for isgri image by Carlo Ferrigno

the original request was through to https://astrooda... .
To retrieve the result, please follow this URL
"

Commit messages can be quite long, multi-line.
I am not sure if URL will comply with restrictions on commit text.

Reply:
> > so, is the user name, extracted from the token, in the commit message ok?
> 
> I think I didn't write this previous message of mine properly and therefore was not clear.

Unclear indeed!

I also deduced from this example:

![image](https://user-images.githubusercontent.com/3909535/161996251-9c1e1d6f-9d7e-4d52-bcb4-35e7233452ee.png)

Gitlab can show distinction between commiter and author, e.g.:

![image](https://user-images.githubusercontent.com/3909535/161996580-ede4a6c5-7daa-4fc4-9483-6cf85313d9e3.png)

We'd want authored-by the token-user and commited-by MMODA.
Authored-by can be set in git commit command. Commited-by will be set automatically.

> 
> > "stored API code of MMODA request for isgri image by Carlo Ferrigno
> 
> > the original request was through to https://astrooda/... .
> > To retrieve the result, please follow this URL"
> 
> Regarding these information to include inside the commit message, some can be extracted from the token (user name) and some from the request. Is this what you mean with metadata that cannot be found in the token?

Information which can not be found in **commit metadata**, not in token. This information comes from token and request, yes. I suppose you have it right there.

> 
> As for the link you mention, that is the link we also include within the email correct?

yes.

Reply:
For example:

![image](https://user-images.githubusercontent.com/3757833/162004114-345d6a66-67ff-461f-ab74-58d9b9d85372.png)

The relative gitlab account is also recognized and set.

Out of curiosity, I also made a test and applied a mock author, that does not have a git lab account:

![image](https://user-images.githubusercontent.com/3757833/162004924-18c0f58a-16cd-4541-a04d-71b2ce10175e.png)

Reply:
Better now:
![image](https://user-images.githubusercontent.com/3757833/162006658-566f7264-a970-4e50-84b6-d760f8ca330d.png)

Reply:
A first commit with the extended message and the url:

![image](https://user-images.githubusercontent.com/3757833/162030923-db13e165-4ef6-4178-99db-1275f9bba75a.png)